### PR TITLE
Lecture 25: Spectral Learning for Graphical Models

### PR DESCRIPTION
**Pull request by xiamengzhou**

### CODE REVIEW

**Comment by zhisbug** [comment]
@xiamengzhou : could you provide a link to the rendered version so that I can review? Thanks!

**Comment by xiamengzhou** [comment]
Sure, here is the link!
https://xiamengzhou.github.io/pgm-spring-2019/notes/lecture-25/

File: _posts/2019-04-17-lecture-25.md [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+  <div class="row">
+    <div class="col two">
+      <img src="{{ 'assets/img/notes/lecture-25/diagram.png' | relative_url }}" />
+    </div>
+  </div>
+</figure>
+
+The problem arises from marginalizing out the latent (white) variables in the diagram - in the graph, we seem to arrive at a clique.
+
+However, if look at a simple latent variable model with one hidden, and 3 observed states depending on that hidden state, we can see that there is a relationship between the number of hidden states for $H$ and the complexity of the relationship between the observed variables once $H$ is margnialized.
```
Comment (zhisbug) [inline _posts/2019-04-17-lecture-25.md]: if look at -> if looking at

File: _posts/2019-04-17-lecture-25.md [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+<figure id="example-figure" class="l-body-outset">
+  <div class="row">
+    <div class="col two">
+      <img src="{{ 'assets/img/notes/lecture-25/hmm-rv.png' | relative_url }}" />
+    </div>
+  </div>
+</figure>
+If we have 1 state, then the marginalized distribution is equal to the conditional distribution of the observed variables on $H$, which means they're all independent. On the other hand if $H$ has $m^3$ states, wher $m$ is number of states for each observed variable, then the marginalization would have enough parameters to fully paramterize any relationship between the observed variables, and thus forming a clique.
```
Comment (zhisbug) [inline _posts/2019-04-17-lecture-25.md]: marginalized -> marginal

Comment (zhisbug) [reply _posts/2019-04-17-lecture-25.md]: Typo: wher?

Comment (zhisbug) [reply _posts/2019-04-17-lecture-25.md]: $m$ is number of ... -> $m$ is the number of ...

File: _posts/2019-04-17-lecture-25.md [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+
+Clearly, there is also an intermediate case between these two extremes - how can we exploit this relationship between observed variables?
+
+### Matrix view of probability rules
```
Comment (zhisbug) [inline _posts/2019-04-17-lecture-25.md]: -> A matrix view of ...

File: _posts/2019-04-17-lecture-25.md [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+
+- Probability form: $$ \mathbb{P}[X] = \sum\limits_Y \mathbb{P}[X\mid Y]\mathbb{P}[Y] $$
+- Matrix algebra form : $$ \mathcal{P}[X] = \mathcal{P}[X\mid Y] \times \mathcal{P}[X] $$ where $\times$ is the matrix dot product.
+
+#### Chain Rule
+
+- Probability form: $$\mathbb{P}[X, Y] = \mathbb{P}[X \mid Y]\mathbb{P}[Y] = \mathbb{P}[Y\mid X]\mathbb{P}[Y]$$
+- Matrix algebra form: $$ \mathcal{P}[X, Y] = \mathcal{P}[X \mid Y] \times \mathcal{P}[\mathbin Y]$$ where $\mathcal{P}[Y]$ is a diagonal matrix w/ the probabilities on the diagonal.
+
+#### Two variable example
```
Comment (zhisbug) [inline _posts/2019-04-17-lecture-25.md]: example -> examples

File: _posts/2019-04-17-lecture-25.md [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+      <img src="{{ 'assets/img/notes/lecture-25/intermediate_case.png' | relative_url }}" />
+    </div>
+  </div>
+</figure>
+
+We can claim that $$rank(\mathcal{P}[A, B]) \leq k$$
+
+**Proof**: We can use the matrix interpretation of the chain rule from above.
+
+We know that $$ \mathcal{P}[A, B] = \mathcal{P}[A\mid X]\mathcal{P}[X]\mathcal{P}[B\mid X]^T $$ We also note that each of these factor matrices have rank $\leq k$, which means that matrix characterizing the joint distribution must have rank $\leq k$.
```
Comment (zhisbug) [inline _posts/2019-04-17-lecture-25.md]: please reformat this sentence as it looks not properly formatted.

File: _posts/2019-04-17-lecture-25.md [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+<figure id="example-figure" class="l-body-outset">
+  <div class="row">
+    <div class="col two">
+      <img src="{{ 'assets/img/notes/lecture-25/intermediate_case_2.png' | relative_url }}" />
+    </div>
+  </div>
+</figure>
+
+Thus, we can see that latent variable models encode *low rank dependencies* among variables, and we can our linear algebra toolkit to exploit the low rank structure.
```
Comment (zhisbug) [inline _posts/2019-04-17-lecture-25.md]: missing a verb?

File: _posts/2019-04-17-lecture-25.md [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+
+## Factorization without latent variable
+
+### How to factorize with latent variables
+
+For the above example, assume $m>k$, we can easily factorize the giant joint distribution table into several small tables by introducing latent variable $H_2$:
+
+$$ \mathcal{P}[X_{\{1,2\}}, X_{\{3,4\}}] = \mathcal{P}[X_{\{1,2\}} | H_2] \mathcal{P}[\oslash H_2]\mathcal{P}[X_{\{3,4\}}|H_2]^T$$
```
Comment (zhisbug) [inline _posts/2019-04-17-lecture-25.md:127]: what does $\oslash$ mean here? better give some context.

File: _posts/2019-04-17-lecture-25.md [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+
+### How to factorize without latent variables
+The goal is to decompose a giant matrix $\mathcal{P}[X_{\{1,2\}}, X_{\{3,4\}}]$ into the product of several small matrices. One observation is that:
+
+$$\mathcal{P}[X_{\{1,2\}}, X_{3}] = \mathcal{P}[X_{\{1,2\}} | H_2] \mathcal{P}[\oslash H_2]\mathcal{P}[X_{3}|H_2]^T$$
+
+$$ \mathcal{P}[X_{2}, X_{\{3,4\}}] = \mathcal{P}[X_{2} | H_2] \mathcal{P}[\oslash H_2]\mathcal{P}[X_{\{3,4\}}|H_2]^T$$
+
+It can be seen that for the terms on the right hand sides of the above equations, 
+$$\mathcal{P}[X_{\{1,2\}} | H_2] \mathcal{P}[\oslash H_2]\mathcal{P}[X_{\{3,4\}}|H_2]^T = \mathcal{P}[X_{\{1,2\}}, X_{\{3,4\}}]$$
```
Comment (zhisbug) [inline _posts/2019-04-17-lecture-25.md]: Align the two equations below?

File: _posts/2019-04-17-lecture-25.md [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+
+$$\mathcal{P}[X_{2} | H_2] \mathcal{P}[\oslash H_2]\mathcal{P}[X_{3}|H_2]^T=\mathcal{P}[X_{2}, X_{3}]$$
+
+which leads to:
+$$\mathcal{P}[X_{\{1,2\}}, X_{\{3,4\}}] = \mathcal{P}[X_{\{1,2\}}, X_{3}]\mathcal{P}[X_{2}, X_{3}]^{-1}\mathcal{P}[X_{2}, X_{\{3,4\}}]$$
+
+which gives a factorization that only contains observed variables, indicating there is no need to learn latent variables using EM, hence more efficient an no local optimal problem. The caveat is that $$\mathcal{P}[X_{2}, X_{3}]$$ have to be invertible. We call this factorization **observable factorization**.
```
Comment (zhisbug) [inline _posts/2019-04-17-lecture-25.md]: -> hence it is more efficient and has no problem of reaching local optima.

File: _posts/2019-04-17-lecture-25.md [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+$$S^{-1}R = \mathcal{P}[X_{2}, X_{3}]^{-1}\mathcal{P}[X_{2}, X_{\{3,4\}}]$$
+Therefore, the observable factorization can be represented as:
+$$M=\mathcal{P}[X_{\{1,2\}}, X_{\{3,4\}}] = \mathcal{P}[X_{\{1,2\}}, X_{3}]\mathcal{P}[X_{2}, X_{3}]^{-1}\mathcal{P}[X_{2}, X_{\{3,4\}}]=LSS^{-1}R$$
+
+## Training / Testing with Spectral learning
+### training
+For observable factorization:
+$$\mathcal{P}[X_{\{1,2\}}, X_{\{3,4\}}] = \mathcal{P}[X_{\{1,2\}}, X_{3}]\mathcal{P}[X_{2}, X_{3}]^{-1}\mathcal{P}[X_{2}, X_{\{3,4\}}]$$
+
+we can simply obtain MLE of the matrices on the right hand side by the counting from samples.
```
Comment (zhisbug) [inline _posts/2019-04-17-lecture-25.md]: -> by counting from samples (remove "the")

File: _posts/2019-04-17-lecture-25.md [hunk 11]
```diff
@@ -0,0 +1,353 @@
...
+\mathcal{P}[X_{\{1, 2\}}, X_3]\mathbf{V}(\mathbf{U}^T \mathcal{P}[X_2, X_3]V)^{-1}U^T \mathcal{P}[X_2, X_{\{3, 4\}}]$$
+
+where $U, V$ are the top left/right k singular vectors of $\mathcal{P}[X_2, X_3]$.
+
+* When $k > m$
+The inverse does exist. But it no longer satisfies the following property, which we used to derive the factorization.
+
+$$ \mathcal{P}(X_2, X_3)^{-1} = (\mathcal{P}[X_3 |H_2]^T)^{-1} \mathcal{P}[\oslash H_2]^{-1} \mathcal{P}[X_2|H_2]^{-1}$$
+
+This is much more difficult to fix, and intuitively corresponds to how the problem becomes intractable if $k >> m$.
```
Comment (zhisbug) [inline _posts/2019-04-17-lecture-25.md]: the condition after "if" is missing.

Comment (zhisbug) [reply _posts/2019-04-17-lecture-25.md]: The indent is missing here. Please fix.

**Review by zhisbug — CHANGES_REQUESTED** [review]
Please address my comments and I'll take a second review.

**Comment by xiamengzhou** [comment]
@zhisbug Please take a look! Resolved all the issues.

**Comment by alshedivat** [comment]
@zhisbug, are we ready to merge?

**Pull request merged**